gadget/device_linux.go

```diff
@@ -137,3 +200,45 @@ func FindMountPointForStructure(ps *PositionedStructure) (string, error) {
 
 	return mountPoint, nil
 }
+
+const writableFsLabel = "writable"
+
+func findParentDeviceWithWritableFallback() (string, error) {
+	byWritableLabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-label/", writableFsLabel)
+	target, err := filepath.EvalSymlinks(byWritableLabel)
+
+	if err != nil {
+		if os.IsNotExist(err) {
+			return "", ErrDeviceNotFound
+		}
+		return "", fmt.Errorf("cannot resolve device symlink for filesystem label %q: %v", writableFsLabel, err)
+	}
+
+	// /dev/sda3 -> sda3
+	devname := filepath.Base(target)
+
+	// do not bother with investigating major/minor devices (inconsistent
+	// across block device types) or mangling strings, but look at sys
+	// hierarchy for block devices instead:
+	// /sys/block/sda               - main SCSI device
+	// /sys/block/sda/sda1          - partition 1
+	// /sys/block/sda/sda<n>        - partition n
+	// /sys/block/nvme0n1           - main NVME device
+	// /sys/block/nvme0n1/nvme0n1p1 - partition 1
+	matches, err := filepath.Glob(filepath.Join(dirs.GlobalRootDir, "/sys/block/*/", devname))
```

Comment:
Looks fine. I was trying to see if there is a simpler and more direct way that doesn't require globbing, but couldn't find it.